[Bug]: Container reuse fails with "The container name X is already in use by container"

### Testcontainers version

0.15.0

### Using the latest Testcontainers version?

Yes

### Host OS

macOS

### Host Arch

arm64

### Go Version

1.18

### Docker version

```shell
Client:
 Cloud integration: v1.0.22
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:46:56 2021
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.5.0 (74594)
 Engine:
  Version:          20.10.12
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.12
  Git commit:       459d0df
  Built:            Mon Dec 13 11:43:07 2021
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```


### Docker info

```shell
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.7.1)
  compose: Docker Compose (Docker Inc., v2.2.3)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 7
  Running: 2
  Paused: 0
  Stopped: 5
 Images: 26
 Server Version: 20.10.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.10.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 5
 Total Memory: 3.841GiB
 Name: docker-desktop
 ID: 66GO:BOPI:CZ4Q:KTVR:3DK6:V2KO:Q32M:73BP:MSQD:UAMO:U2PS:63IG
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5000
  127.0.0.0/8
 Live Restore Enabled: false
```


### What happened?

```
package reusable
​
import (
	"context"
	"fmt"
	"testing"
	"time"
​
	"github.com/docker/go-connections/nat"
	testcontainers "github.com/testcontainers/testcontainers-go"
	"github.com/testcontainers/testcontainers-go/wait"
)
​
const (
	postgresPort     = 5432
	postgresPassword = "test"
	postgresUser     = "test"
	postgresDb       = "test"
)
​
func TestSomething(t *testing.T) {
	t.Parallel()
​
	_, _ = mustSetupPostgresDB(t)
}
​
func TestSomething2(t *testing.T) {
	t.Parallel()
​
	_, _ = mustSetupPostgresDB(t)
}
​
func mustSetupPostgresDB(t *testing.T) (string, testcontainers.Container) {
	natPort := fmt.Sprintf("%d/tcp", postgresPort)
	pg, err := testcontainers.GenericContainer(
		context.Background(),
		testcontainers.GenericContainerRequest{
			ContainerRequest: testcontainers.ContainerRequest{
				Image:        "postgis/postgis",
				Name:         "test-postgres",
				ExposedPorts: []string{natPort},
				Env: map[string]string{
					"POSTGRES_PASSWORD": postgresPassword,
					"POSTGRES_USER":     postgresUser,
					"POSTGRES_DATABASE": postgresDb,
				},
				WaitingFor: wait.ForLog("database system is ready to accept connections").
					WithPollInterval(100 * time.Millisecond).
					WithOccurrence(2),
			},
			Started: true,
			Reuse:   true,
		},
	)
	if err != nil {
		t.Fatal(err)
	}
​
	// Even after log message found Postgres needs a touch more...
	time.Sleep(200 * time.Millisecond)
​
	// Get the container info needed
	port, err := pg.MappedPort(context.Background(), nat.Port(natPort))
	if err != nil {
		t.Fatal(err)
	}
​
	host, err := pg.Host(context.Background())
	if err != nil {
		t.Fatal(err)
	}
​
	return fmt.Sprintf(
		"postgresql://%s:%s@%s:%d/%s?sslmode=disable",
		postgresUser,
		postgresPassword,
		host,
		port.Int(),
		postgresDb,
	), pg
}
```

Attempting to reuse a postgres container between 2 concurrent tests results in an error that the container is already in use.






### Relevant log output

```shell
GOROOT=/opt/homebrew/Cellar/go/1.18.3/libexec #gosetup
GOPATH=/Users/michaelhatch/go #gosetup
/opt/homebrew/Cellar/go/1.18.3/libexec/bin/go test -c -o /private/var/folders/3j/ns56w0zx13nc1504bjrgtqy40000gp/T/GoLand/___1go_test_github_com_nxcr_org_compose.test github.com/nxcr-org/compose #gosetup
/opt/homebrew/Cellar/go/1.18.3/libexec/bin/go tool test2json -t /private/var/folders/3j/ns56w0zx13nc1504bjrgtqy40000gp/T/GoLand/___1go_test_github_com_nxcr_org_compose.test -test.v -test.paniconexit0
=== RUN   TestSomething
=== PAUSE TestSomething
=== CONT  TestSomething
--- PASS: TestSomething (19.90s)
=== RUN   TestSomething2
=== PAUSE TestSomething2
=== CONT  TestSomething2
2022/10/27 11:39:36 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 20.10.12
  API Version: 1.41
  Operating System: Docker Desktop
  Total Memory: 3933 MB
2022/10/27 11:39:36 Starting container id: 1e5ba554050e image: docker.io/testcontainers/ryuk:0.3.4
2022/10/27 11:39:37 Waiting for container id 1e5ba554050e image: docker.io/testcontainers/ryuk:0.3.4
2022/10/27 11:39:37 Container is ready id: 1e5ba554050e image: docker.io/testcontainers/ryuk:0.3.4
    compose_test.go:73: Error response from daemon: Conflict. The container name "/test-postgres" is already in use by container "d0907349f6b47d82d347fb4b3e258d794a7e8c216918aded3a6ab17d09f6d51c". You have to remove (or rename) that container to be able to reuse that name.: failed to create container
--- FAIL: TestSomething2 (0.74s)
2022/10/27 11:39:37 Starting container id: d0907349f6b4 image: postgis/postgis
2022/10/27 11:39:37 Waiting for container id d0907349f6b4 image: postgis/postgis
2022/10/27 11:39:56 Container is ready id: d0907349f6b4 image: postgis/postgis

FAIL

Process finished with the exit code 1
```


### Additional Information

_No response_

--- a/generic.go
+++ b/generic.go
@@ -58,7 +58,7 @@ func GenericContainer(ctx context.Context, req GenericContainerRequest) (Contain
 
 	var c Container
 	if req.Reuse {
-		// we must protect the reusability of the container in the case it's invoke
+		// we must protect the reusability of the container in the case it's invoked
 		// in a parallel execution, via ParallelContainers or t.Parallel()
 		reuseContainerMx.Lock()
 		defer reuseContainerMx.Unlock()

--- a/generic.go
+++ b/generic.go
@@ -4,9 +4,11 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"sync"
 )
 
 var (
+	reuseContainerMx  sync.Mutex
 	ErrReuseEmptyName = errors.New("with reuse option a container name mustn't be empty")
 )
 
@@ -56,6 +58,11 @@ func GenericContainer(ctx context.Context, req GenericContainerRequest) (Contain
 
 	var c Container
 	if req.Reuse {
+		// we must protect the reusability of the container in the case it's invoke
+		// in a parallel execution, via ParallelContainers or t.Parallel()
+		reuseContainerMx.Lock()
+		defer reuseContainerMx.Unlock()
+
 		c, err = provider.ReuseOrCreateContainer(ctx, req.ContainerRequest)
 	} else {
 		c, err = provider.CreateContainer(ctx, req.ContainerRequest)

--- a/parallel_test.go
+++ b/parallel_test.go
@@ -2,9 +2,12 @@ package testcontainers
 
 import (
 	"context"
+	"fmt"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go/wait"
 )
 
 func TestParallelContainers(t *testing.T) {
@@ -118,5 +121,52 @@ func TestParallelContainers(t *testing.T) {
 				t.Fatalf("expected containers: %d, got: %d\n", tc.resLen, len(res))
 			}
 		})
+	}
+}
+
+func TestParallelContainersWithReuse(t *testing.T) {
+	const (
+		postgresPort     = 5432
+		postgresPassword = "test"
+		postgresUser     = "test"
+		postgresDb       = "test"
+	)
+
+	natPort := fmt.Sprintf("%d/tcp", postgresPort)
+
+	req := GenericContainerRequest{
+		ContainerRequest: ContainerRequest{
+			Image:        "postgis/postgis",
+			Name:         "test-postgres",
+			ExposedPorts: []string{natPort},
+			Env: map[string]string{
+				"POSTGRES_PASSWORD": postgresPassword,
+				"POSTGRES_USER":     postgresUser,
+				"POSTGRES_DATABASE": postgresDb,
+			},
+			WaitingFor: wait.ForLog("database system is ready to accept connections").
+				WithPollInterval(100 * time.Millisecond).
+				WithOccurrence(2),
+		},
+		Started: true,
+		Reuse:   true,
+	}
+
+	parallelRequest := ParallelContainerRequest{
+		req,
+		req,
+		req,
+	}
+
+	ctx := context.Background()
+
+	res, err := ParallelContainers(ctx, parallelRequest, ParallelContainersOptions{})
+	if err != nil {
+		e, _ := err.(ParallelContainersError)
+		t.Fatalf("expected errors: %d, got: %d\n", 0, len(e.Errors))
+	}
+
+	for _, c := range res {
+		defer c.Terminate(ctx)
 	}
 }

